Disable the authentication document cache by default, and make it the location of optional UWSGI debugging information.

leonardr opened the pull request:
## Description

This branch productizes two things I've learned while trying to fix SIMPLY-3346:

1. At the moment, the authentication document cache is too much trouble to justify the relatively small performance improvement it brings. I've disabled it on every site NYPL hosts and it should be disabled by default.
2. Some deployment configurations seem to send bad UWSGI data to the app server, but there's no easy way to look at that data on a running site.

I've put item #2 into QA servers as a hotfix; this productizes the hotfix so it can be turned on as necessary, to solve this problem or other problems in the future. If UWSGI debug is enabled (this is a secret site-wide setting called `uwsgi_debug`) you can get UWSGI debug information by making a request to `authentication_document` and putting `debug` in the query string, e.g. `https://qa-circulation.openebooks.us/USOEI/authentication_document?debug`.

## How Has This Been Tested?

In addition to the unit tests I tested this against a local development database.

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

matthewhelt commented on api/controller.py:
```diff
@@ -304,11 +304,14 @@ def get_domain(url):
             ConfigurationSetting.sitewide(
                 self._db, Configuration.AUTHENTICATION_DOCUMENT_CACHE_TIME
-            ).value_or_default(3600)
+            ).value_or_default(0)
```
Is there a way to completely skip the cache instead of setting the retention time span to 0? There might be a performance impact to skip it altogether (although I have no profiled the run time for this). This question is moot if the cache is an intended/required function of the architecture.

leonardr replied:
I could remove the cache altogether without changing much, but conceptually it makes a lot of sense, so I'd like to keep the code in place and fix it (or decide to remove it) later.

matthewhelt commented on tests/test_controller.py:
```diff
@@ -422,7 +425,11 @@ def mock_for_library(cls, incoming_library):
         ConfigurationSetting.sitewide(
             self._db, Configuration.AUTHENTICATION_DOCUMENT_CACHE_TIME
-        ).value = "0"
+        ).value = "60"
```
I'd add a comment that these settings are seconds where the Configuration.AUTHENTICATION_DOCUMENT_CACHE_TIME value is in MS - just for clarity when reviewing test output.

leonardr replied:
Where are you getting the milliseconds idea? The default value for AUTHENTICATION_DOCUMENT_CACHE_TIME is 3600 (which I'm about to change in a follow-up).

matthewhelt reviewed (commented):
Would it not be better to have a separate endpoint for the WSGI data to be populated so one could still evaluate the auth document and also see application state? I think overloading the endpoint with 2 functions may not be ideal if one needed to debug the rendered state of a non-cached auth doc and still be able to grab the state of a given request, this would require a change to log the wsgi request object in the same thread and require more telemetry, so this might be viable for immediate needs regardless.

matthewhelt commented:
> Would it not be better to have a separate endpoint for the WSGI

Its been clarified that this change adds a "_debug" key to the auth document and does not override that document or any content that would be rendered otherwise. As it's not a default option and would be very helpful this presents only the issues of inadvertent use which could disclose information and thus should be documented.

leonardr commented:
I added the docs you requested to the docstring of the controller method. I also changed the term used throughout from "uwsgi" (which is a specific package) to "wsgi" (the name of the spec).

matthewhelt reviewed (approved):
Per approach to adjust timeout in the given architecture.
